src/Directory.Build.targets

```diff
@@ -10,5 +14,24 @@
     <EmbeddedResource Update="**\*.resx" GenerateSource="true" />
   </ItemGroup>
 
+  <ItemGroup>
+    <!-- Workaround for https://github.com/dotnet/project-system/issues/2374 -->
+    <UpToDateCheckInput Include="$(CodeAnalysisRuleSet)" Condition="'$(CodeAnalysisRuleSet)' != ''" />
+
+    <!-- Include the shipping rules explicitly when they are imported by the non-shipping rules -->
+    <!-- Additional workaround for https://github.com/dotnet/project-system/issues/2374 -->
+    <UpToDateCheckInput Include="$(MSBuildThisFileDirectory)..\build\Analyzers_ShippingRules.ruleset" Condition="'$(IsTestProject)' == 'true' OR '$(NonShipping)' == 'true'" />
+  </ItemGroup>
+
+  <PropertyGroup>
+    <!-- Workaround for https://github.com/dotnet/roslyn/issues/25041 -->
+    <MSBuildAllProjects Condition="'$(IsTestProject)' == 'true' OR '$(NonShipping)' == 'true'">$(MSBuildAllProjects);$(MSBuildThisFileDirectory)..\build\Analyzers_ShippingRules.ruleset</MSBuildAllProjects>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- Workaround for https://github.com/xunit/xunit/issues/1651 -->
+    <None Remove="$(NuGetPackageRoot)xunit.runner.visualstudio\$(XUnitRunnerVisualStudioVersion)\build\net20\..\_common\xunit.abstractions.dll" />
+  </ItemGroup>
```

Comment:
There is a similar work-around in the toolset: https://github.com/dotnet/roslyn-tools/blob/7d7c6f32462ab4575b0f8db1f760a4483ea41c4f/sdks/RepoToolset/tools/Workarounds.targets#L38-L54

Is it incorrect? Perhaps it should be updated instead of adding another one here.

Reply:
➡️ That workaround is not correct for two reasons:

1. It operates on `Content` instead of `None`, which changed in xunit/visualstudio.xunit@11c164546b017a9a666e941561e355e58ebd2371
2. It's wrapped inside a `<Target>`, which means the exclusion won't play into the fast up-to-date check in Visual Studio

Hopefully xunit addresses the issue (xunit/xunit#1651) and eliminates the need for workarounds. In the meantime, this change is a big improvement to build times for this repository and even if another project adopts a workaround it will not cause a problem.

Reply:
I submitted dotnet/roslyn-tools#165 as well